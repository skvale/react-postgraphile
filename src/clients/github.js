import { ApolloClient } from 'apollo-client'
import { createHttpLink } from 'apollo-link-http'
import { setContext } from 'apollo-link-context'
import { InMemoryCache } from 'apollo-cache-inmemory'

const authLink = auth => setContext((_, { headers }) => {
  return {
    headers: {
      ...headers,
      authorization: auth ? `Bearer ${auth}` : null,
    }
  }
})

const httpLink = createHttpLink({
  uri: 'https://api.github.com/graphql',
})

export default auth => new ApolloClient({
  link: authLink(auth).concat(httpLink),
  cache: new InMemoryCache()
})
