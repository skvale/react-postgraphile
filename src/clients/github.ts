import graphql from 'graphql.js'

export default (token: string) =>
  graphql('https://api.github.com/graphql', {
    headers: {
      'Access-Token': token
    }
  })
