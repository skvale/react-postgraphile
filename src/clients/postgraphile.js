import { ApolloClient } from "apollo-client";
import { createHttpLink } from "apollo-link-http";
import { setContext } from "apollo-link-context";
import { InMemoryCache } from "apollo-cache-inmemory";

const uri = "http://localhost:5000/graphql";

const httpLink = createHttpLink({ uri });

const authLink = token =>
  setContext((_, { headers }) => ({
    headers: {
      ...headers,
      ...(token ? { Authorization: `Bearer ${token}` } : {})
    }
  }));

export default token => {
  return new ApolloClient({
    link: authLink(token).concat(httpLink),
    cache: new InMemoryCache()
  });
};

export const noAuthClient = new ApolloClient({
  link: httpLink,
  cache: new InMemoryCache()
});
