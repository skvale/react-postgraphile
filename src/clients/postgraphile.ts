import graphql from 'graphql.js'
import Cookies from 'js-cookie'

const uri = 'http://localhost:5000/graphql'

export const createClient = (token = Cookies.get('postgraphile-jwt')) =>
  graphql(uri, {
    asJSON: true,
    headers: token
      ? {
          Authorization: `Bearer ${token}`
        }
      : {}
  })
