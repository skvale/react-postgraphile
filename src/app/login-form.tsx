import React from 'react'
import { useMachine } from '@xstate/react'
import { print } from 'graphql'
import Cookies from 'js-cookie'
import { useQueryParam, BooleanParam } from 'use-query-params'
import { loader } from 'graphql.macro'
import { Button } from '../components/button'
import { Input } from '../components/input'
import { Form } from '../components/form'
import { Person } from '../schema'
import { Link } from '../components/link'
import { Card } from '../components/card'
import {
  loginMachine,
  registerMachine,
  LoginContext
} from '../machines/login-machine'
import { createClient } from '../clients/postgraphile'

const registerPersonMutation = loader('./graphql/register-person-mutation.gql')
const authenticateMutation = loader('./graphql/authenticate-mutation.gql')

const login = (...args: any[]) =>
  createClient()(print(authenticateMutation))(...args)
const register = (...args: any[]) =>
  createClient()(print(registerPersonMutation))(...args)

export type LoginFormProps = {
  currentUser?: Person | null
  updateToken: (jwtToken: string) => void
}

enum ViewState {
  LOGIN = 'login',
  REGISTER = 'register'
}

export const LoginForm: React.FC<LoginFormProps> = ({
  currentUser,
  updateToken
}) => {
  const [registerParam] = useQueryParam('register', BooleanParam)

  const [current, send] = useMachine(
    registerParam ? registerMachine : loginMachine,
    {
      services: {
        onLogin: async ({ email, password }: LoginContext) => {
          const result = await login({
            email,
            password
          })
          console.log(result.authenticate)
          if (result && result.authenticate && result.authenticate.jwtToken) {
            Cookies.set('postgraphile-jwt', result.authenticate.jwtToken)
            updateToken(result.authenticate.jwtToken)
            return result.authenticate.jwtToken
          } else {
            throw Error('something happened')
          }
        },
        onRegister: async ({
          email,
          password,
          firstName,
          lastName
        }: LoginContext) => {
          return register({
            email,
            firstName,
            lastName,
            password
          })
        }
      }
    }
  ) as any

  const { error, firstName, email, lastName, password } = current.context

  if (current.matches('success')) {
    return null
  }

  const onSwapState = () => {
    send('TOGGLE')
  }

  return (
    <div className='flex justify-center'>
      {currentUser && <div>Currently logged in as: {currentUser.fullName}</div>}
      <Card
        className='bg-white'
        title={current.matches(ViewState.LOGIN) ? 'Login' : 'Register'}
      >
        <div className='float-right'>
          {current.matches(ViewState.LOGIN) ? (
            <Link data-testid='toggle-link' color='gray' onClick={onSwapState}>
              Register
            </Link>
          ) : (
            <Link data-testid='toggle-link' color='gray' onClick={onSwapState}>
              Login
            </Link>
          )}
        </div>
        <Form
          data-testid='login-form'
          error={error}
          onSubmit={(e: React.FormEvent<HTMLFormElement>) => {
            send({
              type: 'SUBMIT'
            })
          }}
        >
          <div className='mb-6'>
            <Input
              data-testid='login-email'
              id='email'
              label='Email'
              onChange={e => send({ type: 'EMAIL', value: e.target.value })}
              value={email}
            />
            <Input
              data-testid='login-password'
              id='password'
              label='Password'
              type='password'
              onChange={e => send({ type: 'PASSWORD', value: e.target.value })}
              value={password}
            />
            {current.matches(ViewState.REGISTER) && (
              <React.Fragment>
                <Input
                  data-testid='login-first-name'
                  id='firstName'
                  label='First Name'
                  onChange={e =>
                    send({ type: 'FIRST_NAME', value: e.target.value })
                  }
                  value={firstName}
                />
                <Input
                  data-testid='login-last-name'
                  id='lastName'
                  label='Last Name'
                  onChange={e =>
                    send({ type: 'LAST_NAME', value: e.target.value })
                  }
                  value={lastName}
                />
              </React.Fragment>
            )}
          </div>
          <div className='flex items-center justify-between'>
            <Button data-testid='login-submit-button' type='submit'>
              {current.matches(ViewState.LOGIN) ? 'Sign In' : 'Register'}
            </Button>
          </div>
        </Form>
      </Card>
    </div>
  )
}
