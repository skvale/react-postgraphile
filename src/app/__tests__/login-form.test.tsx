import '@testing-library/jest-dom/extend-expect'

import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { LoginForm, LoginFormProps } from '../login-form'
import * as graph from '../../clients/postgraphile'

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve))
}

let defaultProps: LoginFormProps
beforeEach(() => {
  defaultProps = {
    updateToken: jest.fn()
  }
})

test('renders', async () => {
  const { asFragment } = render(<LoginForm {...defaultProps} />)
  expect(asFragment()).toMatchSnapshot()
})

test('click login', async () => {
  const login = jest.fn(() =>
    Promise.resolve({ authenticate: { jwtToken: 'mock-jwt-token' } })
  )
  // @ts-ignore
  graph.createClient = () => () => login
  const { findByTestId } = render(<LoginForm {...defaultProps} />)
  const email = await findByTestId('login-email')
  const password = await findByTestId('login-password')
  fireEvent.change(email, { target: { value: 'foo@bar.com' } })
  fireEvent.change(password, { target: { value: 'password' } })
  const form = await findByTestId('login-form')
  fireEvent.submit(form)
  expect(login).toBeCalledTimes(1)
  expect(login.mock.calls[0]).toMatchSnapshot()
  await flushPromises()
  expect(defaultProps.updateToken).toHaveBeenCalledWith('mock-jwt-token')
})

test('renders registration view', async () => {
  const { asFragment, findByText } = render(<LoginForm {...defaultProps} />)
  const registerLink = await findByText('Register')
  fireEvent.click(registerLink)
  expect(asFragment()).toMatchSnapshot()
})

test('click register', async () => {
  const register = jest.fn()
  // @ts-ignore
  graph.createClient = () => () => register
  const { findByTestId } = render(<LoginForm {...defaultProps} />)
  const registerLink = await findByTestId('toggle-link')
  fireEvent.click(registerLink)
  const email = await findByTestId('login-email')
  const password = await findByTestId('login-password')
  const firstName = await findByTestId('login-first-name')
  const lastName = await findByTestId('login-last-name')
  fireEvent.change(email, { target: { value: 'foo@bar.com' } })
  fireEvent.change(password, { target: { value: 'password' } })
  fireEvent.change(firstName, { target: { value: 'First' } })
  fireEvent.change(lastName, { target: { value: 'Last' } })
  const form = await findByTestId('login-form')
  fireEvent.submit(form)
  expect(register).toBeCalledTimes(1)
  expect(register.mock.calls[0][0]).toMatchSnapshot()
})
