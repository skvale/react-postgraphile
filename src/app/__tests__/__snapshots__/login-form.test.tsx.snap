// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`click login 1`] = `
Array [
  Object {
    "email": "foo@bar.com",
    "password": "password",
  },
]
`;

exports[`click register 1`] = `
Object {
  "email": "foo@bar.com",
  "firstName": "First",
  "lastName": "Last",
  "password": "password",
}
`;

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center"
  >
    <div
      class="max-w-sm rounded overflow-hidden shadow-lg bg-white"
    >
      <div
        class="p-6"
      >
        <div
          class="font-bold text-xl mb-2"
        >
          Login
        </div>
        <div
          class="text-gray-700 text-base"
        >
          <div
            class="float-right"
          >
            <a
              class="cursor-pointer font-bold inline-block align-baseline text-sm text-blue-700 hover:text-blue-900"
              color="gray"
              data-testid="toggle-link"
            >
              Register
            </a>
          </div>
          <form
            class="bg-white pt-6 pb-4"
            data-testid="login-form"
          >
            <div
              class="mb-6"
            >
              <div
                class="mb-4"
              >
                <label
                  class="block text-gray-700 text-sm font-bold mb-2"
                  for="email"
                >
                  Email
                </label>
                <input
                  class="appearance-none border leading-tight py-2 px-3 rounded shadow text-gray-700 w-full focus:outline-none focus:shadow-outline"
                  data-testid="login-email"
                  id="email"
                  value=""
                />
              </div>
              <div
                class="mb-4"
              >
                <label
                  class="block text-gray-700 text-sm font-bold mb-2"
                  for="password"
                >
                  Password
                </label>
                <input
                  class="appearance-none border leading-tight py-2 px-3 rounded shadow text-gray-700 w-full focus:outline-none focus:shadow-outline"
                  data-testid="login-password"
                  id="password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="flex items-center justify-between"
            >
              <button
                class="bg-blue-500 hover:bg-blue-700 text-white py-2 px-4 rounded focus:outline-none focus:shadow-outline"
                data-testid="login-submit-button"
                type="submit"
              >
                Sign In
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders registration view 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center"
  >
    <div
      class="max-w-sm rounded overflow-hidden shadow-lg bg-white"
    >
      <div
        class="p-6"
      >
        <div
          class="font-bold text-xl mb-2"
        >
          Register
        </div>
        <div
          class="text-gray-700 text-base"
        >
          <div
            class="float-right"
          >
            <a
              class="cursor-pointer font-bold inline-block align-baseline text-sm text-blue-700 hover:text-blue-900"
              color="gray"
              data-testid="toggle-link"
            >
              Login
            </a>
          </div>
          <form
            class="bg-white pt-6 pb-4"
            data-testid="login-form"
          >
            <div
              class="mb-6"
            >
              <div
                class="mb-4"
              >
                <label
                  class="block text-gray-700 text-sm font-bold mb-2"
                  for="email"
                >
                  Email
                </label>
                <input
                  class="appearance-none border leading-tight py-2 px-3 rounded shadow text-gray-700 w-full focus:outline-none focus:shadow-outline"
                  data-testid="login-email"
                  id="email"
                  value=""
                />
              </div>
              <div
                class="mb-4"
              >
                <label
                  class="block text-gray-700 text-sm font-bold mb-2"
                  for="password"
                >
                  Password
                </label>
                <input
                  class="appearance-none border leading-tight py-2 px-3 rounded shadow text-gray-700 w-full focus:outline-none focus:shadow-outline"
                  data-testid="login-password"
                  id="password"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="mb-4"
              >
                <label
                  class="block text-gray-700 text-sm font-bold mb-2"
                  for="firstName"
                >
                  First Name
                </label>
                <input
                  class="appearance-none border leading-tight py-2 px-3 rounded shadow text-gray-700 w-full focus:outline-none focus:shadow-outline"
                  data-testid="login-first-name"
                  id="firstName"
                  value=""
                />
              </div>
              <div
                class="mb-4"
              >
                <label
                  class="block text-gray-700 text-sm font-bold mb-2"
                  for="lastName"
                >
                  Last Name
                </label>
                <input
                  class="appearance-none border leading-tight py-2 px-3 rounded shadow text-gray-700 w-full focus:outline-none focus:shadow-outline"
                  data-testid="login-last-name"
                  id="lastName"
                  value=""
                />
              </div>
            </div>
            <div
              class="flex items-center justify-between"
            >
              <button
                class="bg-blue-500 hover:bg-blue-700 text-white py-2 px-4 rounded focus:outline-none focus:shadow-outline"
                data-testid="login-submit-button"
                type="submit"
              >
                Register
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
